Add text length limit on TextBox control

okonk opened the pull request:
As per the discussion, went with an approach that once the limit is reached no more text can be entered. This seems to be the standard behaviour used by other applications and makes sense.

Can adjust naming/stuff if required.

okonk committed:
Add a length limit to textboxes. Also basic testing controls.

billyquith commented on source/gwork/include/Gwork/Controls/TextBox.h:
```diff
@@ -63,6 +63,9 @@ namespace Gwk
             virtual void SetCursorEnd(int i);
 
+            void SetMaxTextLength(int maxLength) { m_maxTextLength = maxLength; }
+            const int& GetMaxTextLength() const { return m_maxTextLength; }
```
We can return `int` here

billyquith commented on source/gwork/source/Controls/TextBox.cpp:
```diff
@@ -46,6 +46,7 @@ GWK_CONTROL_CONSTRUCTOR(TextBox)
     m_cursorLine = 0;
     m_bEditable = true;
     m_bSelectAll = false;
+    m_maxTextLength = -1;
```
Be nice if `-1` wasn't a magic number. Perhaps `static constexpr NO_MAX_LENGTH = -1;` ?

billyquith commented on source/test/source/api/TextBox.cpp:
```diff
@@ -68,6 +81,13 @@ class TextBox : public TestUnit
+        {
+            Gwk::Controls::TextBoxMultiline* label = new Gwk::Controls::TextBoxMultiline(this);
+            label->SetText("Max Length 20");
+            label->SetMaxTextLength(20);
```
After this, perhaps you could set the text to something longer than the maximum length to test whether this works as expected?

Grant1219 reviewed (commented):
I would prefer to use `size_t` when dealing with string lengths, but I realize `int` is being used here for the special case where there is no text limit and `m_maxTextLength` is set to `-1`.

billyquith commented:
> I would prefer to use `size_t` when dealing with string length

Fair comment, but `int` is used throughout the code. I have a ticket #45 to work on this.

billyquith reviewed (changes requested):
Could you please make the changes suggested. I'm more keen on the demo that the change works, and return type.

okonk commented:
What happens is the text will be larger (since we're only checking the limit on insert), but nothing bad happens, the user just has to take the text below the limit before they can edit it. That seems alright to me. I'll add a test for it anyway

okonk committed:
Suggested changes:
Changed return type to int.
Added a constant to replace magic number.
Added a test covering longer than max length.

billyquith commented:
Yep, but I think you added another control, so just want to distinguish what the control is demonstrating. E.g. setText(...), setMaxTextLen(6), setText("Maximum text length"), so we can see that the length has been clamped/truncated, and that any editing maintains this. - possibly you could add a counter next to it showing the text length, but that is entirely optional.

Grant1219 commented on source/gwork/include/Gwork/Controls/TextBox.h:
```diff
@@ -90,6 +90,8 @@ namespace Gwk
             Event::Caller onTextChanged;
             Event::Caller onReturnPressed;
 
+            static constexpr int NO_MAX_LENGTH = -1;
```
This could just be `static const int NO_MAX_LENGTH = -1;`

billyquith replied:
This project is C++11 so there is no reason not to use `constexpr`.

okonk committed:
Remove extra TextBoxMultiline test, add SetText past length limit to previous test

okonk commented:
Alright, changed it to add an extra SetText after SetMaxTextLength

billyquith commented on source/gwork/source/Controls/TextBox.cpp:
```diff
@@ -80,7 +80,7 @@ void TextBox::InsertText(const Gwk::String& strInsert)
 
     int insertSize = static_cast<int>(strInsert.size());
-    if (m_maxTextLength > -1 && TextLength() + insertSize > m_maxTextLength)
+    if (m_maxTextLength > NO_MAX_LENGTH && TextLength() + insertSize > m_maxTextLength)
```
This should be != NO_MAX_LENGTH as it is a magic number out of the usual range.

billyquith reviewed (approved):
Thanks for making the changes. I haven't tested them but look good.
